rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,775 @@
+
+    def _features_as_token_ids(
+        self, features: List[Union[tf.Tensor, tf.SparseTensor]]
+    ) -> Optional[tf.Tensor]:
+        """Creates dense labels (token IDs) used for negative sampling in MLM."""
+        # If there are dense features, we use them as labels - taking the first dense
+        # feature in the list because any dense feature will do the job.
+        for f in features:
+            if not isinstance(f, tf.SparseTensor):
+                return tf.stop_gradient(f)
+
+        # If no dense features are found, use a sparse feature but convert it into
+        # a dense one first.
+        for f in features:
+            if isinstance(f, tf.SparseTensor):
+                return tf.stop_gradient(self._tf_layers["sparse_to_dense_token_ids"](f))
+
+    def _create_mlm_variables(
+        self,
+        sequence_features: List[Union[tf.Tensor, tf.SparseTensor]],
+        seq_sent_features: tf.Tensor,
+        mask_combined_sequence_sentence: tf.Tensor,
+        training: Union[tf.Tensor, bool],
+    ):
+        """Produce helper variables for masked language modelling (only in training).
+        
+        The `token_ids` embeddings can be viewed as token-level labels/unique IDs of all
+        input tokens (to be used later in the MLM loss) because these embeddings aren't 
+        affected by dropout or masking and are effectively always unique for different 
+        input tokens (and same for the same tokens).
+        `token_ids` share the batch and sequence dimension with the combined sequence-
+        and sentence-level features, the last dimension is unimportant and mimics the 
+        first dense sequence-level feature in the list of features, or alternatively the
+        last dimension will have size 2 if there are only sparse sequence features
+        present.
+        """
+        token_ids = self._features_as_token_ids(sequence_features)
+        # Pad in the sequence dimension to match the shape of combined sequence- and
+        # sentence-level features (sentence-level features effectively have sequence
+        # length of 1).
+        token_ids = tf.pad(token_ids, [[0, 0], [0, 1], [0, 0]])
+
+        # mlm_mask_bool has the same shape as mask_combined_sequence_sentence (i.e. as
+        # the tensor with all combined features), with True meaning tokens that are
+        # masked and False meaning tokens that aren't masked or that are fake
+        # (padded) tokens.
+        transformer_inputs, mlm_mask_bool = self._tf_layers["mlm_input_mask"](
+            seq_sent_features, mask_combined_sequence_sentence, training
+        )
+        return transformer_inputs, token_ids, mlm_mask_bool
+
+    def call(
+        self,
+        inputs: Tuple[
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            tf.Tensor,
+            tf.Tensor,
+        ],
+        training: Optional[Union[tf.Tensor, bool]] = None,
```

Comment:
since these are rasa layers, I think you can require for `training` to be `bool`.
```suggestion
        training: bool = False,
```
Otherwise, you need to substitute all the `if` with it with `smart_cond` from `tf`

Reply:
Hmmmm, didn't think about this at all. Thanks!